### PR TITLE
mobile: Add `rules_android_ndk`

### PR DESCRIPTION
**Pull request by phlax**

### CODE REVIEW

**Comment by phlax** [comment]
cc @keith i had a go at adding `rules_android_ndk` again - but without updating the ndk - it fails with foreign_cc issues - any ideas why?

**Comment by phlax** [comment]
fail:

```console
CMakeFiles/cmTC_2eff2.dir/testCCompiler.c.o
    [2/2] Linking C executable cmTC_2eff2
    FAILED: cmTC_2eff2 
    : && /mnt/engflow/worker/work/1/exec/external/android_ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -no-canonical-prefixes --target=x86_64-linux-android21 -fdiagnostics-color -Wa,--noexecstack -faddrsig -faligned-new -fdata-sections -ffunction-sections -funsigned-char -fstack-protector -g -funwind-tables -DNDEBUG -fno-strict-aliasing -fomit-frame-pointer -g0 -Oz -Wall -Wformat-security -Wno-char-subscripts -Wno-error=deprecated-declarations -Wno-maybe-uninitialized -Wno-sign-compare -Wno-strict-overflow -Wno-unused-but-set-variable -Wunused-but-set-parameter -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-ignored-optimization-argument -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-error=deprecated-declarations -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-implicit-int-float-conversion -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-private-header -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unused-const-variable -Wno-unused-function -Wno-unused-private-field -Wno-user-defined-warnings -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-designator -Wno-gnu-variable-sized-type-not-at-end -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-tautological-constant-in-range-compare -Wno-unknown-pragmas -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wgnu-redeclared-enum -Winfinite-recursion -Wliteral-conversion -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wunused-comparison -Wvla -Wthread-safety-analysis -fPIC -DABSL_MIN_LOG_LEVEL=4 -fdebug-types-section -Wno-deprecated-declarations -Wno-deprecated-declarations -Os -fvisibility=hidden -fexceptions -no-canonical-prefixes --target=x86_64-linux-android21 -fdiagnostics-color -Wl,-z,noexecstack -Wl,-zseparate-code -Wl,--no-rosegment -Wl,--gc-sections -Wl,--exclude-libs,libgcc.a -Wl,--build-id=md5 -Wl,--undefined-glob='Java_*' -Wl,--undefined-glob='JNI_*' -pie -Wl,--exclude-libs,libunwind.a -fuse-ld=lld -Wl,--icf=safe -Wl,--no-undefined -lc -lm -latomic -ldl CMakeFiles/cmTC_2eff2.dir/testCCompiler.c.o -o cmTC_2eff2   && :
    ld.lld: error: unknown -z value: separate-code
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    ninja: build stopped: subcommand failed.
    ```

**Comment by keith** [comment]
yea i think so

File: mobile/bazel/BUILD [hunk 1]
```diff
@@ -25,6 +25,12 @@ alias(
     visibility = ["//library/jni:__pkg__"],
 )
 
+alias(
```
Comment (abeyad) [inline mobile/bazel/BUILD:28]: what is the purpose of this target?

Comment (phlax) [reply mobile/bazel/BUILD:28]: envoy/mobile has some formatting rule that says this is the only place you can refer to `@bazel_tools`

File: mobile/bazel/rules_android_ndk.patch [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+-                    features = ["crosstool_linker_lld"],
+-                ),
+                 # C++ compiles
+                 flag_set(
+                     actions = actions.all_cpp_compile,
+@@ -1381,7 +1376,7 @@ def ndk_cc_toolchain_config(
+                 ),
+                 flag_set(
+                     actions = actions.all_link,
+-                    flags = ["-lc", "-lm", "-latomic", "-ldl"],
```
[the remark this answers is not in the document]
Comment (phlax) [reply mobile/bazel/rules_android_ndk.patch:21]: fails with the compiler we have with our current ndk

its not a case of any more - we didnt have these rules until now - previously we used the native (deprecated) rules

**Review by RyanTheOptimist — APPROVED** [review]
I don't really understand this very well, but I'm happy to defer to Ali. If he's happy, then I'm happy.

**Review by abeyad — APPROVED** [review]

**Pull request merged**